improve TypeScript declarations

xuoe opened the pull request:
Relates to #7.

Commit:
improve TypeScript declarations

Relates to #7.

xuoe commented on index.d.ts:
```diff
@@ -1,17 +1,20 @@
+type Options = {
   skipInitialCall?: boolean;
   equals?: typeof shallowEquals;
 }
```
What would be the benefit of an interface here?

KagamiChan replied:
basically type and interface make not much difference here, interface might fit object typing better

xuoe commented on index.d.ts:
```diff
@@ -1,17 +1,20 @@
   equals?: typeof shallowEquals;
 }
 
-type Observer<S = any, MS = any> = unknown;
```
The  `observer()` "constructor" function (defined below) returns a function that takes `(state, dispatch, options)`. The function itself was not intended to be called by client code, but I think a more descriptive type alias would be helpful here. Moreover, allowing the user to call it would not cause any issues.

I've elided `options` from the signature as I personally find it a bit confusing to leave them in. Those options would be considered `globals` (those passed in to `observe()`). Anyway, `options` could be accommodated, too, if need be.

xuoe commented on index.d.ts:
```diff
@@ -1,17 +1,20 @@
-type Observer<S = any, MS = any> = unknown;
-type Mapper<S = any, MS = any> = (state: S) => MS;
-type Dispatcher<MS> = (dispatch: Dispatch<MS>, currentState: MS, previousState: MS) => unknown;
+type Mapper<S, MS> = (state: S) => MS;
```
I'm not too sure this is correct. A `mapper` always takes a `store` state and can return any value. Is there a way to enforce `S` to be a POJO? Is it necessary?

I feel that `S = any, MS = any` is too loose and insufficiently descriptive. I may be wrong, though. (First time writing any TypeScript here.)

(By the way, `MS` stands for "MappedState")

KagamiChan replied:
`state` is not necessarily an object if some one really wants that way, so we might leave it

KagamiChan replied:
state is not necessarily an object if someone really wants, we might leave it to be loose

xuoe commented on index.d.ts:
```diff
@@ -1,17 +1,20 @@
-type Dispatcher<MS> = (dispatch: Dispatch<MS>, currentState: MS, previousState: MS) => unknown;
+type Mapper<S, MS> = (state: S) => MS;
+type Observer<S = any> = (state: S, dispatch: Dispatch<AnyAction>) => void;
+type Dispatcher<MS> = (dispatch: Dispatch<AnyAction>, current: MS, previous?: MS) => void;
```
`Dispatch<AnyAction>` works with both Redux `3.7.2` and `4.0.4`. I think `Dispatch<MS>` is incorrect in the previous commit. Also, both the `Observer` and `Dispatcher` return nothing.

`state` is mandatory here, so shouldn't `Observer<S = any>` actually be `Observer<S>`? Would that break the `observe()` function, which takes an observer array (`Observer[]`) as one of its arguments?

KagamiChan replied:
To explain, Dispatcher<MS> should be meaningful for `@types/redux`, which is a third-party maintained type and differ from the current official typing (and that's why we need some updates here)

xuoe commented on index.d.ts:
```diff
@@ -1,17 +1,20 @@
-export function observer<S, MS>(mapper: Mapper<S, MS>, dispatcher: Dispatcher<MS>, options?: Options): Observer<S, MS>;
-export function observer<S>(dispatcher: Dispatcher<S>, options?: Options): Observer<S, S>;
+interface ObserverCreator {
+    <S>(dispatcher: Dispatcher<S>, locals?: Options): Observer<S>
```
`Dispatcher` is defined as `Dispatcher<MS>` above, but here we're using `<S>` instead, which I would assume to be the `store` state and not a mapped state (which is what the dispatcher is called with at runtime). Is this important or am I reading too much into it?

xuoe commented on index.d.ts:
```diff
@@ -1,17 +1,20 @@
+}
 
-export function observe<S>(store: Store<S>, observers: Observer[], options?: Options): Unsubscribe;
+export const observer: ObserverCreator
```
Capturing the different `observer()` signatures under `ObserverCreator` seems a bit more helpful. Not sure if `ObserverCreator` should be exported as well?

KagamiChan commented on index.d.ts:
```diff
@@ -1,17 +1,20 @@
 export function shallowEquals(a: any, b: any): boolean;
 
-interface Options {
+type Options = {
```
we might export `Options` too

Commit:
add TypeScript declaration tests

This commit ensures the TypeScript declarations introduced recently work
with all relevant redux versions (>3.4).

Commit:
enfore npm >6.9 during development

This ensures package aliases work as expected.

See https://github.com/npm/cli/releases/v6.9.0 for more details.

xuoe commented:
I've added tests for redux `3.4`, `3.7` and `>4` inside the `types/` directory (the files are more or less the same, except for the redux version). I've mocked the interfaces provided by redux in `types/redux.d.ts`, which are used internally by the exported types/interfaces declared in `types/index.d.ts`.

@ankon, @KagamiChan, if you have the time, let me know if there's anything missing/you'd like to add before I push the changes to npm. Thanks!
